proto/spacemesh/global_state_types.proto

```diff
@@ -0,0 +1,71 @@
+syntax = "proto3";
+package spacemesh;
+import "spacemesh/types.proto";
+
+// Data items related to an account
+enum AccountDataFlag {
+    NONE = 0;
+    TRANSACTION_RECIEPT = 1; // tx receipt for a tx to or from an account
+    REWARD = 2; // reward awarded to an account
+}
+
+message AccountDataParams {
+    AccountId account_id = 1;
+    uint32 account_data_bit_field = 2; // A bit field of AccountDataFlag
+}
+
+message AccountDataQueryParams {
+    AccountDataParams account_data = 1;
+    uint32 max_results = 2; // max numbers of results client would like to get
+    uint32 offset = 3; // return results from offset
+}
+
+// An account related data item
+message AccountData {
+    oneof item {
+        Reward reward = 1;
+        TransactionReceipt receipt = 2;
+        Account account = 3;
```

Comment:
Why not separate out the various event types encapsulated in an update to Account? E.g., NewNonce, NewBalance, NewCode, etc.

Also, it sort of bothers me that the event types contained in this oneof/enum actually overlap: e.g., a `Reward` event will correspond to a `NewBalance` event. Is there some more intuitive way to bundle these?

Reply:
Good idea but it requires more work on the server-side - fire an event every time a field changes instead of firing one event after these change. It is going to be very common that counter and balance change - every time there's a tx from that account. I think it is okay by now and moves the responsibility for clients to update their data with the new account data. I don't think strongly about this but I think it is good enough for the first version.

Reply:
Changed to DataItem and cleaned up the types a bit.

Reply:
I'm still confused by this message. I'd expect to see a stream of all events that originate from or touch an account: updates to nonce and balance, rewards, transaction state changes for sent transactions, and maybe a list of transactions for incoming transactions. I don't understand what the 'Account' message is doing in this case. It seems to be a convenience wrapper around nonce + balance changes, but I'd rather include those as first class data items here. Also incoming transactions seems to be missing.

Reply:
Are you sure we want wallets to get 2 separate events when counter and balance change? it will be very common that these change in the same time when there was an outgoing tx processed.

Regarding transactions results changes you comment about - this is provided the tx receipt event type and it includes incoming coin transfer txs and outgoing transactions.

Reply:
Also - reward and new balance are not the same - new balance will just show a new balance while reward will show the actual reward and tx fee assigned by an stf in a layer.

Reply:
Maybe my discomfort with this the way it stands here is that it mixes two concerns: _cause_ and _effect._ The _cause_ of a change in the account state is the tx or the reward. The _effect_ is the updated balance/nonce. I can think of two ways to improve this:

1. Bundle cause and effect together into a single 'event' type that always includes both, e.g.,

```
message AccountUpdatedEvent {
    oneof item {
        Reward reward = 1;
        TransactionReceipt receipt = 2;
    }
    Account updated_account_data = 3;
}
```

2. Or else, split this into two different streams and event types, e.g.,

```
message AccountEvent {
    oneof item {
        Reward reward = 1;
        TransactionReceipt receipt = 2;
    }
}
message AccountUpdatedEvent {
    Account updated_account_data = 1;
}
```

Does this make sense?

Reply:
not really. The way we use this pattern now is that the client specifies what types of events it wants and he can specify he only cares about rewards and/or tx receipts but not in account updates. I really don't see a reason to change it. Rewards and receipts are side-effects and account data change is a change in the account db. I think that the proposed change just over-complicate the api where it is not really needed. If you specify a bitmap of Reward || Account then you should get back both. If you only specify you want to see rewards then you just specify Rewards in the bitmap. In both cases we should give the caller what he asked for.

Reply:
So the current design is simpler and gives more flexibility to callers.

Reply:
A similar pattern is now in GlobalStateStream - we don't want to force clients to get account data when they specified they are not interested in this data in the stream.

Reply:
We discussed this on a call just now. I'm still a little uncomfortable with this mix of concerns, but I'm okay with it in the interest of keeping the API simpler, of multiplexing the streams (#40), and as long as it's clearly documented. I documented this more clearly in the comments.